File: simp-front-end/src/lib/api/cauldron/cauldron.ts
import axios from "axios";
import {
  Cauldron,
  CauldronCreate,
  CauldronUpdate,
  CauldronData,
  CauldronDataCreate,
  CauldronDataUpdate,
  CauldronRecipe,
} from "@/lib/types/cauldron";

const API_BASE_URL = process.env.RECIPES_API_URL || "http://localhost:8020/v1";

const apiClient = axios.create({
  baseURL: API_BASE_URL,
  withCredentials: true,
});

// Create a new cauldron record (to add a recipe to the user's cauldron)
export async function createCauldron(
  data: CauldronCreate
): Promise<Cauldron> {
  const response = await apiClient.post<Cauldron>("/cauldrons", data);
  return response.data;
}

// Update an existing cauldron record
export async function updateCauldron(
  cauldronId: string,
  data: CauldronUpdate
): Promise<Cauldron> {
  const response = await apiClient.put<Cauldron>(`/cauldrons/${cauldronId}`, data);
  return response.data;
}

// Delete a cauldron record by its ID (to remove a recipe from the user's cauldron)
export async function deleteCauldron(
  cauldronId: string
): Promise<{ detail: string }> {
  const response = await apiClient.delete<{ detail: string }>(`/cauldrons/${cauldronId}`);
  return response.data;
}

// Retrieve paginated cauldron records for a specific user
export async function getCauldronsByUser(
  userId: string,
  skip = 0,
  limit = 10
): Promise<{ cauldrons: Cauldron[]; total: number }> {
  const response = await apiClient.get<{ cauldrons: Cauldron[]; total: number }>(`/cauldrons/user/${userId}`, {
    params: { skip, limit },
  });
  return response.data;
}

// Retrieve paginated cauldron recipes (combined cauldron and recipe data) for a specific user
export async function getCauldronRecipes(
  userId: string,
  skip = 0,
  limit = 10
): Promise<{ cauldron_recipes: CauldronRecipe[]; total_cauldron_recipes: number }> {
  const response = await apiClient.get<{
    cauldron_recipes: CauldronRecipe[];
    total_cauldron_recipes: number;
  }>(`/cauldrons/recipes`, {
    params: { user_id: userId, skip, limit },
  });
  return response.data;
}

// Delete a cauldron record by user id and recipe id.
// This endpoint finds the cauldron entry based on the user and recipe IDs.
export async function deleteCauldronByUserAndRecipe(
  userId: string,
  recipeId: string
): Promise<{ detail: string }> {
  const response = await apiClient.delete<{ detail: string }>(`/cauldrons/user/${userId}/recipe/${recipeId}`);
  return response.data;
}

// Optionally, you can add similar API calls for CauldronData if needed.

File: simp-api-recipes/api/cauldron.py
from typing import Any, Dict, List, Union
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func
from sqlalchemy.orm import Session
import uuid

from database.connection import SessionLocal
from schemas.cauldron import CauldronCreate, CauldronSchema, CauldronUpdate
from models.database_tables import Cauldron as CauldronModel  # SQLAlchemy model for Cauldron
from models.database_tables import Recipe

router = APIRouter(tags=["cauldrons"])

def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

@router.post("/", response_model=CauldronSchema)
def create_cauldron(cauldron_create: CauldronCreate, db: Session = Depends(get_db)):
    """
    Create a new cauldron entry to add a recipe to the user's cauldron.
    """
    # Optionally verify the recipe exists or check for duplicates here.
    new_cauldron = CauldronModel(
        user_id=cauldron_create.user_id,
        recipe_id=cauldron_create.recipe_id,
        is_active=cauldron_create.is_active,
    )
    db.add(new_cauldron)
    db.commit()
    db.refresh(new_cauldron)
    return new_cauldron

@router.delete("/{cauldron_id}", response_model=Dict[str, str])
def delete_cauldron(cauldron_id: uuid.UUID, db: Session = Depends(get_db)):
    """
    Delete a cauldron entry, effectively removing the recipe from the user's cauldron.
    """
    cauldron_obj = (
        db.query(CauldronModel)
        .filter(CauldronModel.cauldron_id == cauldron_id)
        .first()
    )
    if not cauldron_obj:
        raise HTTPException(status_code=404, detail="Cauldron not found")
    
    db.delete(cauldron_obj)
    db.commit()
    return {"detail": "Cauldron deleted successfully"}

@router.put("/{cauldron_id}", response_model=CauldronSchema)
def update_cauldron(
    cauldron_id: uuid.UUID,
    cauldron_update: CauldronUpdate,
    db: Session = Depends(get_db)
):
    """
    Update a cauldron entry. For example, you might change the is_active flag.
    """
    cauldron_obj = (
        db.query(CauldronModel)
        .filter(CauldronModel.cauldron_id == cauldron_id)
        .first()
    )
    if not cauldron_obj:
        raise HTTPException(status_code=404, detail="Cauldron not found")
    
    update_data = cauldron_update.dict(exclude_unset=True)
    for key, value in update_data.items():
        setattr(cauldron_obj, key, value)
    
    db.commit()
    db.refresh(cauldron_obj)
    return cauldron_obj

@router.get("/user/{user_id}", response_model=Dict[str, Union[List[CauldronSchema], int]])
def read_cauldrons_by_user(
    user_id: uuid.UUID,
    skip: int = Query(0, ge=0),
    limit: int = Query(10, ge=1, le=100),
    db: Session = Depends(get_db)
):
    """
    Retrieve paginated cauldron entries for a user.
    """
    total_cauldrons = (
        db.query(CauldronModel)
        .filter(CauldronModel.user_id == user_id)
        .count()
    )

    cauldrons = (
        db.query(CauldronModel)
        .filter(CauldronModel.user_id == user_id)
        .offset(skip)
        .limit(limit)
        .all()
    )

    return {
        "cauldrons": cauldrons,
        "total": total_cauldrons,
    }

@router.get("/recipes", response_model=Dict[str, Any])
def read_cauldron_recipes(
    user_id: uuid.UUID,
    skip: int = Query(0, ge=0),
    limit: int = Query(10, ge=1, le=100),
    db: Session = Depends(get_db)
):
    """
    Retrieve cauldron recipes (i.e. recipes added to the cauldron) for a given user.
    This endpoint joins the cauldron entries with their corresponding recipe data.
    """
    total = (
        db.query(func.count(CauldronModel.cauldron_id))
        .filter(CauldronModel.user_id == user_id)
        .scalar()
    )

    results = (
        db.query(CauldronModel, Recipe)
        .join(Recipe, CauldronModel.recipe_id == Recipe.recipe_id)
        .filter(CauldronModel.user_id == user_id)
        .offset(skip)
        .limit(limit)
        .all()
    )

    cauldron_recipes = []
    for cauldron, recipe in results:
        cauldron_recipes.append({
            "cauldron_id": str(cauldron.cauldron_id),
            "user_id": str(cauldron.user_id),
            "recipe_id": str(cauldron.recipe_id),
            "is_active": cauldron.is_active,
            "created_at": cauldron.created_at.isoformat(),
            "updated_at": cauldron.updated_at.isoformat(),
            "title": recipe.title,
            "tags": [tag.name for tag in recipe.tags],
            "front_image": recipe.front_image,
        })

    return {
        "cauldron_recipes": cauldron_recipes,
        "total_cauldron_recipes": total,
    }

@router.delete("/user/{user_id}/recipe/{recipe_id}", response_model=Dict[str, str])
def delete_cauldron_by_user_and_recipe(
    user_id: uuid.UUID,
    recipe_id: uuid.UUID,
    db: Session = Depends(get_db)
):
    """
    Delete a cauldron entry for a specific user and recipe.
    """
    cauldron_obj = (
        db.query(CauldronModel)
        .filter(
            CauldronModel.user_id == user_id,
            CauldronModel.recipe_id == recipe_id
        )
        .first()
    )
    if not cauldron_obj:
        raise HTTPException(
            status_code=404,
            detail="Cauldron entry not found for the given user and recipe"
        )
    db.delete(cauldron_obj)
    db.commit()
    return {"detail": "Cauldron entry deleted successfully"}
